### PR TITLE
Create infrastructure to separate Foundation into multiple modules; move all code into FoundationBase module.

### PR DESCRIPTION
**Pull request by millenomi**
This commit moves all of {Swift,}Foundation into a separate module named {Swift,}FoundationSoil, including the Core Foundation symbols, and then creates a new {Swift,}Foundation module that re-exports Soil so that `import {Swift,}Foundation` clients can continue working correctly. **The module name isn’t final; do not use it.**

Future split work can define multiple modules this way by copying the approach in both the Xcode project and the CMake setup.

This requires a coordinated clang change (forthcoming).

### CODE REVIEW

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245

@swift-ci please test

**Comment by millenomi** [comment]
cc @gottesmm @compnerd

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914

@swift-ci please test

File: Foundation/Foundation.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+//
+// Copyright (c) 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+
+import CoreFoundation
+
+#if DEPLOYMENT_RUNTIME_SWIFT && canImport(Darwin)
```
Comment (compnerd) [inline Foundation/Foundation.swift]: Isn't this due to the ObjC Foundation existing and not specific to Darwin?  Can we not do something like `@_runtime(_ObjC)` here? :-(

Comment (millenomi) [reply Foundation/Foundation.swift]: I’ve tried really hard to confine private attributes elsewhere, but it is correct that the one you point is the most relevant condition.

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Unsure why the clang patch isn't being picked up.

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
The macOS crash is suspect, but unclear if patch-related.

**Comment by millenomi** [comment]
Rebased onto master to ensure this isn't hitting a since-fixed bug.

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
I don't see this crash without my patch; building on macOS to check.

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by millenomi** [comment]
Please test with the following:
https://github.com/apple/swift-clang/pull/245
https://github.com/apple/swift-package-manager/pull/1914
https://github.com/apple/swift/pull/21234

@swift-ci please test

**Comment by shahmishal** [comment]
The Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please re-create pull request with main branch.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412

**Pull request closed without merging**